File: Celebratorica/app/src/main/java/io/blushine/celebratorica/item/Item.java
package io.blushine.celebratorica.item;

import android.support.annotation.NonNull;

import java.text.ParsePosition;
import java.text.SimpleDateFormat;
import java.util.Date;

import io.blushine.android.common.DateFormats;

/**
 * A celebration item
 */
class Item implements Comparable<Item> {
private final SimpleDateFormat DATE_FORMAT = getDateFormat();
private long mItemId = -1;
private long mCategoryId = -1;
private String mText = "";
private String mDate = "";
private long mDateTime = -1;

/**
 * @return date format used for displaying celebrations
 */
static SimpleDateFormat getDateFormat() {
	return DateFormats.getMediumDateFormat();
}

long getItemId() {
	return mItemId;
}

void setItemId(long itemId) {
	mItemId = itemId;
}

long getCategoryId() {
	return mCategoryId;
}

void setCategoryId(long categoryId) {
	mCategoryId = categoryId;
}

String getText() {
	return mText;
}

void setText(String text) {
	mText = text;
}

String getDate() {
	return mDate;
}

void setDate(String date) {
	mDate = date;
	ParsePosition parsePosition = new ParsePosition(0);
	mDateTime = DATE_FORMAT.parse(mDate, parsePosition).getTime();
}

void setDate(long dateTime) {
	mDateTime = dateTime;
	mDate = DATE_FORMAT.format(new Date(dateTime));
}

long getDateTime() {
	return mDateTime;
}

@Override
public int compareTo(@NonNull Item other) {
	// This is before
	if (mDateTime < other.mDateTime) {
		return -1;
	}
	// Other is before
	else if (mDateTime > other.mDateTime) {
		return 1;
	}
	// Else date is same, sort by id
	else if (mItemId < other.mItemId) {
		return -1;
	} else {
		return 1;
	}
}

@Override
public int hashCode() {
	int result = (int) (mItemId ^ (mItemId >>> 32));
	result = 31 * result + (int) (mCategoryId ^ (mCategoryId >>> 32));
	return result;
}

@Override
public boolean equals(Object o) {
	if (this == o) {
		return true;
	}
	if (o == null || getClass() != o.getClass()) {
		return false;
	}
	
	Item that = (Item) o;
	
	return mItemId == that.mItemId && mCategoryId == that.mCategoryId;
	
}
}

File: Celebratorica/app/src/main/java/io/blushine/rmw/item/Item.kt
package io.blushine.rmw.item

import io.blushine.android.common.DateFormats
import java.text.ParsePosition
import java.text.SimpleDateFormat
import java.util.*

/**
 * A celebration item
 */
internal class Item : Comparable<Item> {
	var userId = ""
	var id = ""
	var categoryId = ""
	var text = ""
	var date: Long = -1

	/**
	 * @return The date in human readable format
	 */
	fun getDateString(): String {
		return DATE_FORMAT.format(Date(date))
	}

	/**
	 * Set the date from a string
	 */
	fun setDate(dateString: String) {
		val parsePosition = ParsePosition(0)
		date = DATE_FORMAT.parse(dateString, parsePosition).time
	}

	override fun compareTo(other: Item): Int {
		return java.lang.Long.compare(date, other.date)
	}

	override fun hashCode(): Int {
		return Objects.hash(id)
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) {
			return true
		}
		if (other == null || javaClass != other.javaClass) {
			return false
		}

		val that = other as Item?

		return id == that!!.id

	}

	companion object {
		val DATE_FORMAT: SimpleDateFormat
			get() = DateFormats.getMediumDateFormat()
	}
}
